llvmPackage_17.mlir: cmake files are broken

### Steps To Reproduce

Steps to reproduce the behavior:
1. Build a package that depends on mlir via cmake and get the following error:
```console
CMake Error at /nix/store/zjjkxmbpwxh6w8f3gg145k9zg1wvl432-mlir-17.0.6-dev/lib/cmake/mlir/MLIRTargets.cmake:2017 (message):
  The imported target "MLIRPresburger" references the file

     "/nix/store/zjjkxmbpwxh6w8f3gg145k9zg1wvl432-mlir-17.0.6-dev/lib/libMLIRPresburger.a"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/nix/store/zjjkxmbpwxh6w8f3gg145k9zg1wvl432-mlir-17.0.6-dev/lib/cmake/mlir/MLIRTargets.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /nix/store/zjjkxmbpwxh6w8f3gg145k9zg1wvl432-mlir-17.0.6-dev/lib/cmake/mlir/MLIRConfig.cmake:36 (include)
```

Now it wasn't quite obvious to me at first, but I have a theory what is happening here:
The the package has an `out` and `dev` output, thus the static libraries end up in the `out` output, while the cmake files end up in the `dev` output. I don't really know how output splitting works, but I guess that nix simply replaces all references to the `$out` path, for each output pretending there's only a single such out path, therefore creating a path where the static libraries don't actually exists.

This is further confirmed looking at the store paths of the out output:
- which mlir-tblgen
/nix/store/8qd7dxd0skkf6hqmfhqb5k9r3hgh61mp-mlir-17.0.6/bin/mlir-tblgen
- ls /nix/store/8qd7dxd0skkf6hqmfhqb5k9r3hgh61mp-mlir-17.0.6/lib/libMLIRPresburger.a
/nix/store/8qd7dxd0skkf6hqmfhqb5k9r3hgh61mp-mlir-17.0.6/lib/libMLIRPresburger.a
(out output store path + path from the cmake output above)
- ls /nix/store/8qd7dxd0skkf6hqmfhqb5k9r3hgh61mp-mlir-17.0.6/lib/cmake
ls: cannot access '/nix/store/8qd7dxd0skkf6hqmfhqb5k9r3hgh61mp-mlir-17.0.6/lib/cmake': No such file or directory
(no cmake in the out output store path)
- ls /nix/store/zjjkxmbpwxh6w8f3gg145k9zg1wvl432-mlir-17.0.6-dev/lib/
cmake
(the dev output contains only cmake in the lib directory)


### Additional context

I've built mlir locally with `outputs = ["out"]`, and the error is gone.

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@lxsameer @dtzWill @Ericson2314 @lovek323 @alyssais @RaitoBezarius @rrbutani @sternenseemann 

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/compilers/llvm/common/mlir/default.nix
+++ b/pkgs/development/compilers/llvm/common/mlir/default.nix
@@ -1,6 +1,9 @@
-{ lib, stdenv, llvm_meta
+{ lib
+, stdenv
+, llvm_meta
 , buildLlvmTools
-, monorepoSrc, runCommand
+, monorepoSrc
+, runCommand
 , cmake
 , ninja
 , libxml2
@@ -14,7 +17,7 @@ stdenv.mkDerivation rec {
   inherit version doCheck;
 
   # Blank llvm dir just so relative path works
-  src = runCommand "${pname}-src-${version}" {} ''
+  src = runCommand "${pname}-src-${version}" { } ''
     mkdir -p "$out"
     cp -r ${monorepoSrc}/cmake "$out"
     cp -r ${monorepoSrc}/mlir "$out"
@@ -45,7 +48,7 @@ stdenv.mkDerivation rec {
     "-DLLVM_INSTALL_TOOLCHAIN_ONLY=OFF"
     "-DMLIR_TOOLS_INSTALL_DIR=${placeholder "out"}/bin/"
     "-DLLVM_ENABLE_IDE=OFF"
-    "-DLLD_INSTALL_PACKAGE_DIR=${placeholder "out"}/lib/cmake/mlir"
+    "-DMLIR_INSTALL_PACKAGE_DIR=${placeholder "out"}/lib/cmake/mlir"
     "-DLLVM_BUILD_TESTS=${if doCheck then "ON" else "OFF"}"
     "-DLLVM_ENABLE_FFI=ON"
     "-DLLVM_HOST_TRIPLE=${stdenv.hostPlatform.config}"

